src/stores/redis.rs

```diff
@@ -0,0 +1,436 @@
+
+        match self.seconds {
+            None => con
+                .set::<String, String, String>(generated_key, serde_json::to_string(&val).unwrap())
+                .unwrap(),
+            Some(s) => con
+                .set_ex::<String, String, String>(
+                    generated_key,
+                    serde_json::to_string(&val).unwrap(),
+                    s as usize,
+                )
+                .unwrap(),
+        };
+    }
+}
+
+impl<K, V> Default for RedisCache<K, V>
+where
+    K: Display,
+    V: Serialize + DeserializeOwned + Clone,
+{
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<'de, K, V> Cached<K, V> for RedisCache<K, V>
+where
+    K: Display,
+    V: Serialize + DeserializeOwned + Clone,
+{
+    fn cache_get(&mut self, key: &K) -> Option<&V> {
+        self.clean_up();
+
+        match self.base_get(key) {
+            Some(val) => {
+                self.hits += 1;
+                Some(self.base_vec_fill(&val))
+            }
+            None => {
+                self.misses += 1;
+                None
+            }
+        }
+    }
+
+    fn cache_get_mut(&mut self, key: &K) -> Option<&mut V> {
+        self.clean_up();
+
+        match self.base_get(key) {
+            Some(val) => {
+                self.hits += 1;
+                Some(self.base_vec_fill(&val))
+            }
+            None => {
+                self.misses += 1;
+                None
+            }
+        }
+    }
+
+    fn cache_set(&mut self, key: K, val: V) -> Option<V> {
+        self.clean_up();
+        let old_val = self
+            .base_get(&key)
+            .map(|val| serde_json::from_str(&val).unwrap());
+        self.base_set(&key, &val);
+        old_val
+    }
+
+    fn cache_get_or_set_with<F: FnOnce() -> V>(&mut self, key: K, f: F) -> &mut V {
+        self.clean_up();
+        match self.base_get(&key) {
+            Some(val) => {
+                self.hits += 1;
+                self.base_vec_fill(&val)
+            }
+            None => {
+                self.misses += 1;
+                let val = f();
+                self.base_set(&key, &val);
+                let index = self.store.len();
+                self.store.push(val);
+                &mut self.store[index]
+            }
+        }
+    }
+
+    fn cache_remove(&mut self, key: &K) -> Option<V> {
+        self.clean_up();
+        let mut con = self.client.get_connection().unwrap();
+        self.base_get(&key).map(|val| {
+            con.del::<String, ()>(self.generate_key(key)).unwrap();
+            serde_json::from_str(&val).unwrap()
+        })
+    }
+
+    fn cache_clear(&mut self) {
+        self.clean_up();
+        let mut con = self.client.get_connection().unwrap();
+        redis_rs::cmd("EVAL")
+            .arg("local keys = redis.call('keys', ARGV[1]) \n for i=1,#keys,5000 do \n redis.call('del', unpack(keys, i, math.min(i+4999, #keys))) \n end \n return keys")
```

Comment:
some comments on what's going on here would be helpful. Is there a benefit of doing this over regular code?

Reply:
I'll put a comment in the code also.
But it's the way to delete an unlimited number of keys, matching with a pattern, in Redis :|

Copied from here: https://stackoverflow.com/questions/4006324/how-to-atomically-delete-keys-matching-a-pattern-using-redis#comment39607023_16974060